### ISSUE TITLE
userns=host behavior different from non-namespaced with CMD and ENTRYPOINT statements

### ISSUE THREAD

**Issue by domdom82** [docker/docker#27775]
**Description**
Not sure if DUP, sorry.

I am running user namespaces on Docker 1.12.1 and want to disable them for certain containers.

I have noticed different behavior when launching a container that has a CMD and ENTRYPOINT set in the Dockerfile vs. passing the same commands as `docker run` arguments.

Consider Zookeeper:3.4

https://github.com/31z4/zookeeper-docker/blob/7e7eac6d6c11428849ec13bb7d240e4cfa21b2e7/3.4.9/Dockerfile

Has the following settings:

```
ENTRYPOINT ["/docker-entrypoint.sh"]
CMD ["zkServer.sh", "start-foreground"]
```

When running this on a namespaced daemon (i.e. `--userns-remap=default`) but with `--userns=host` you get:

```
docker run --userns host zookeeper:3.4
/docker-entrypoint.sh: line 15: /conf/zoo.cfg: Permission denied
```

(internally, Zookeeper uses a "zookeeper" user that was created in the Dockerfile and can now not read the /conf/zoo.cfg)

However, if you override the ENTRYPOINT and CMD statements by passing them on the CLI directly, it works fine:

```
docker run --userns host zookeeper:3.4 /docker-entrypoint.sh zkServer.sh start-foreground
ZooKeeper JMX enabled by default
Using config: /conf/zoo.cfg
(...)
```

**Steps to reproduce the issue:**
1. `dockerd --userns-remap=default`
2. `docker pull zookeeper:3.4`
3. `docker run --userns=host zookeeper:3.4`

**Describe the results you received:**
```
/docker-entrypoint.sh: line 15: /conf/zoo.cfg: Permission denied
```
**Describe the results you expected:**
```
ZooKeeper JMX enabled by default
Using config: /conf/zoo.cfg
(...)
```

**Additional information you deem important (e.g. issue happens only occasionally):**
Haven't tried yet, but should also apply to other images with similar properties.

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
Containers: 28
 Running: 1
 Paused: 0
 Stopped: 27
Images: 128
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/165536.165536/aufs
 Backing Filesystem: extfs
 Dirs: 174
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: host bridge null overlay
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.17-boot2docker
Operating System: Boot2Docker 1.12.1 (TCL 7.2); HEAD : ef7d0b4 - Thu Aug 18 21:18:06 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 995.9 MiB
Name: whisk
ID: 2GG5:Y3O4:OYJW:PYDV:NN3G:AYU5:VJZD:XZU3:MFIN:LNNK:YWQW:GVGR
Docker Root Dir: /mnt/sda1/var/lib/docker/165536.165536
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 21
 Goroutines: 38
 System Time: 2016-10-26T09:20:34.541171929Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
Insecure Registries:
 127.0.0.0/8
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
Using docker-machine.

**Comment by justincormack** [comment]
@domdom82 I don't think this has been reported before. What is the `uid` of the user zookeeper creates?

@estesp any ideas?

**Comment by domdom82** [comment]
@justincormack the uid of the zookeeper user is 1000 in the image.
interestingly enough, I found a deviation in the mapping:

The "zookeeper" user is created in the Dockerfile, so it pre-exists when the container is launched.
But when I `su` from root to the user, the uid on the host is unchanged:

```
docker run  -it zookeeper:3.4 sh
/zookeeper-3.4.9 # su - zookeeper
```

Output of `ps -xauf` on the host:

```
root     17929  0.1  5.1 424084 52644 ?        Sl   Oct25   2:06 dockerd -D -g /var/lib/docker -H unix
root     17935  0.0  1.5 256304 15776 ?        Ssl  Oct25   0:02  \_ docker-containerd -l unix:///var/
root     31039  0.0  0.2 185456  2892 ?        Sl   12:11   0:00      \_ docker-containerd-shim c3c9e9
165536   31049  0.2  0.0   1524     4 pts/1    Ss   12:11   0:00          \_ sh
166536   31068  0.0  0.0   1520     4 pts/1    S+   12:11   0:00              \_ -sh
```

My `/etc/subuid` shows

```
dockremap:165536:65536
```

I would expect the second shell to run as uid 166536 + 1000 (since zookeeper has uid 1000) but it doesn't. The shell starts with the same uid as remapped root on the host.

However, if I create a user interactively in the container using

```
/zookeeper-3.4.9 # adduser testuser
su - testuser
```

I get a different `ps -xauf` output on the host:

```
root     17929  0.1  5.1 424084 52824 ?        Sl   Oct25   2:06 dockerd -D -g /var/lib/docker -H unix
root     17935  0.0  1.5 256304 15776 ?        Ssl  Oct25   0:02  \_ docker-containerd -l unix:///var/
root     31172  0.0  0.4 185456  4936 ?        Sl   12:15   0:00      \_ docker-containerd-shim 5aeddd
165536   31183  0.0  0.0   1524     4 pts/1    Ss   12:15   0:00          \_ sh
166537   31210  0.0  0.0   1520     4 pts/1    S+   12:16   0:00              \_ -sh
```

This is still not what I would expect (testuser had uid 1001 in the container, should be 166536 + 1001) but at least it is not the remapped root user.

So to me it seems pre-existing users in the container are ignored for the user namespace mappings..?

**Comment by estesp** [comment]
I may need to try and reproduce with the boot2docker image, as I am actually getting the expected results on a standard Ubuntu 16.04 system with Docker 1.12.x

For example, if I `su-exec zookeeper` in the image, I get a process running as {userns-base} + 1000 (the zookeeper user created in the Dockerfile):
```
231072   13624  0.0  0.0   6220  1804 pts/4    Ss   15:57   0:00 bash
232072   13653  0.0  0.0   6216  1904 pts/4    S+   15:57   0:00 bash
```

(231072 is my remapped root; so that is the shell executed by my `docker run`; 232072 is the zookeeper user used via `su-exec zookeeper bash`)

I also don't have the permissions issue running with `--userns=host` as expected if the function is working properly. I'm concerned something is not working properly in the boot2docker image related to user namespaces or user id mapping?

**Comment by domdom82** [comment]
@estesp I can confirm the uid mapping problem with userns enabled does not appear on Ubuntu (I used 14.04).

However, for me the permissions issue still persists.

I think the main problem should be this:

1. If you configure your daemon for user namespace mapping, a folder `/var/lib/docker/<uid>.<gid>` is created. All images that you pull and all container fs you create are now owned by the remapped uid/gid.
2. This usually works, if you run your image with userns enabled and do a `ls -lha /` on the image you get:
  ```
drwxr-xr-x    2 root     root        4.0K Oct 19 00:39 bin
drwxr-xr-x    2 zookeepe zookeepe    4.0K Oct 26 14:18 conf
drwxr-xr-x    2 zookeepe zookeepe    4.0K Oct 26 14:18 data
drwxr-xr-x    2 zookeepe zookeepe    4.0K Oct 26 14:18 datalog
drwxr-xr-x    5 root     root         380 Oct 26 14:18 dev
-rwxrwxr-x    1 root     root         873 Oct 19 00:40 docker-entrypoint.sh
drwxr-xr-x   22 root     root        4.0K Oct 26 14:18 etc
(...)
  ```
So directories are correctly owned by the zookeeper user as defined in the image.

3. If you run the same image with `--userns=host` it simply skips the remapping but the files are **still** owned by the remapped user:

  ```
drwxr-xr-x    2 165536   165536      4.0K Oct 19 00:39 bin
drwxr-xr-x    2 166536   166536      4.0K Oct 26 14:22 conf
drwxr-xr-x    2 166536   166536      4.0K Oct 26 14:22 data
drwxr-xr-x    2 166536   166536      4.0K Oct 26 14:22 datalog
drwxr-xr-x    5 root     root         380 Oct 26 14:22 dev
-rwxrwxr-x    1 165536   165536       873 Oct 19 00:40 docker-entrypoint.sh
drwxr-xr-x   22 165536   165536      4.0K Oct 26 14:22 etc
(...)
  ```

This becomes a problem if you need to run something as a non-root user. i.e. the "zookeeper" user still exists but it does no longer own the files and directories. Btw. I encountered the same problem with the Kafka image which also uses its own "kafka" user.

**Comment by estesp** [comment]
Ah, you are correct--I didn't go far enough with your example to try and "be" the non-root user when in a `--userns=host` executed shell and then run a script that tries to write to the config (like the echo value > zk-config section)

We were well aware of the disconnect of the filesystem ownership when running `--userns=host` but considered it to be a reasonable deficiency for the use cases we were aware of (the need for privileged "admin-like" containers on a daemon that is running user namespaced-containers). Long term we want Linux kernel filesystem support to mount the layers with "shifting" handled by the filesystem user namespace-aware code; until then we have this requirement to have a local cache which is fixed at the remapped range.

Will have to think about what we can do, if anything, to make the non-root `--userns=host` path not painful. The initial ideas that come to mind are scarily hacky for now :-]

**Comment by domdom82** [comment]
I see, so this is a limitation in part of the kernel. I should be able to work around it for the time being by reducing `--userns=host` cases to a minimum. My original idea was to go the path of least change to the original deployment by only using user namespaces where necessary and keeping the pre-existing way otherwise. Turns out I have to invert that logic now :-)

Thanks for your input!

**Comment by matthewdfuller** [comment]
I believe I have a related case. I am running the [Zookeeper/Exhibitor container](https://github.com/mbabineau/docker-zk-exhibitor) which has an `ENTRYPOINT` as `ENTRYPOINT ["bash", "-ex", "/opt/exhibitor/wrapper.sh"]`. To run with user namespaces enabled (via global `DOCKEROPTS`), I am doing:

```
DOCKER_OPTS=--userns-remap=default
```

```
docker run --name=zk-exhibitor --privileged --userns=host -p 8181:8181 -p 2181:2181 -p 2888:2888 -p 3888:3888 -e ZK_ENSEMBLE_SIZE=3 -e ZK_PASSWORD=password -e HOSTNAME=10.1.2.3 mbabineau/zookeeper-exhibitor
```

However, this causes an issue inside of the container. The `/opt/zookeeper/conf/zoo.cfg` file, which is created at runtime by the [Java run command](https://github.com/mbabineau/docker-zk-exhibitor/blob/master/include/wrapper.sh#L90), is owned by the root container user and is not accessible for some reason.

```
/opt/zookeeper/conf# ls -la
total 20
drwxr-xr-x.  2 101000 101000   88 Nov 11 21:29 .
drwxr-xr-x. 12 100000 100000 4096 Nov 11 16:30 ..
-rw-rw-r--.  1 101000 101000  535 Feb 20  2014 configuration.xsl
-rw-rw-r--.  1 101000 101000 2161 Feb 20  2014 log4j.properties
-rw-r--r--.  1 root   root    348 Nov 11 21:29 zoo.cfg
-rw-rw-r--.  1 101000 101000  922 Feb 20  2014 zoo_sample.cfg
```

```
ERROR com.netflix.exhibitor.core.activity.ActivityLog  ZooKeeper Server: JMX enabled by default [pool-3-thread-1]
ERROR com.netflix.exhibitor.core.activity.ActivityLog  ZooKeeper Server: Using config: /opt/zookeeper/bin/../conf/zoo.cfg [pool-3-thread-1]
INFO  com.netflix.exhibitor.core.activity.ActivityLog  ZooKeeper Server: Starting zookeeper ... 2016-11-11 21:29:25,751 [myid:] - INFO  [main:QuorumPeerConfig@103] - Reading configuration from: /opt/zookeeper/bin/../conf/zoo.cfg [pool-3-thread-2]
```

It's unable to read/use the zoo.cfg file, and thus, behaves as if it is in standalone mode rather than respecting the global config.

I haven't found a way around this, but am looking for suggestions.

**Comment by domdom82** [comment]
@matthewdfuller yep this is the same issue. I have decided to run zookeeper in user namespace mode instead and it works fine so far. 

so if you drop the `--userns=host` param it should work for you. however you also need to drop `--privileged` (is there a reason why you need to run privileged? should not be necessary usually)

**Comment by Robotex** [comment]
Has this issue received any update? I believe I'm encountering this same problem.
I have different machines running Ubuntu 16 and Debian 8, all configured with --userns-remap=default and running the same Dockerfile running with --userns=host and rootless user inside. Those on Ubuntu 16 are running without an issue, however those on Debian 8 are facing Permission denied errors in almost every directory including the user's home.

**Comment by svengo** [comment]
The It doesn't make the problem any better, but the behavior is described in the documentation: [Isolate containers with a user namespace](https://docs.docker.com/engine/security/userns-remap/#disable-namespace-remapping-for-a-container).
Does anybody have a workaround?
